### DISCUSSION TITLE
Ella Kaye - Advent of Code 2023 wrap up

### DISCUSSION THREAD

**Discussion by giscus** [EllaKaye/ellakaye-quarto#10] [General]
# Ella Kaye - Advent of Code 2023 wrap up

Reflections on participating in this seasonal coding challenge.

https://ellakaye.co.uk/posts/2023-12-30_advent-of-code-2023/

**Comment by jprowle** [comment]
Thanks for a great resource! Long after the event I did come up with a solution to Day 24 Part 2 in base R based on solving the system of equations.  All the work is in rearranging the equations. Must sign up for the R leaderboard next year.

Day 24 Part 2 Maths:

Calculation of a system of equations

$x=qx[i]t+px[i] y=qy[i]t+py[i] z=qz[i]t+pz[i]$ Equations for hailstone I (p are postions q velocities)

$x=At+X0 y=Bt+Y0 z=Zt+Z0$ Equations for the Rock

Equate x, y and z values and then rearrange in t

$t=(X0-px[i])/(qx[i]-A)=(Y0-py[i])/(qy[i]-B)=(Z0-pz[i])/(qz[i]-C)$

We only need to solve for x and y to start with

$(X0-px[i]).(qy[i]-B)=(Y0-py[i]).(qx[i]-A)$

$qy[i].X0-px[i].qy[i]-X0.B+px[i].B=qx[i].Y0-py[i].qx[i]-Y0.A+py[i].A$

From the above equation, subtract same equation but for hailstone j - This eliminates the quadratic terms X0.B & Y0.A

$qy[j].X0-px[j].qy[j]-X0.B+px[j].B=qx[j].Y0-py[j].qx[j]-Y0.A+py[j].A$

Rearranges to the final linear equation in 4 variables

$X0(qy[i]-qy[j])+B(px[i]-px[j])+Y0(qx[i]-qx[j])+A(py[i]-py[j])=px[i].qy[i]-px[j].qy[j]]-py[i].qx[i]+py[j].qx[j]$

Can use data from 4 or 5 hailstones to generate 4 simultaneous linear equations suitable for the solve function

Then to solve for z now knowing A, B, X0 & Y0

$(Z0-pz[i])/(qz[i]-C)=(X0-px[i])/(qx[i]-A)$

Linear equation in 2 variables use 2 hailstones to make 2 equations and solve

$Z0+C(X0-px[i])/(qx[i]-A)=pz[i]+qz[i].(X0-px[i])/(qx[i]-A)$ 

#Code for part 2 hailstone positions in dataframe p and velocities in dataframe q


AA=NULL
BB=NULL
for(i in 1:4){
  j=i+1   #uses 5 hailstones but 4 would be enough for this method
  AA=rbind(AA,matrix(c((q$y[i]-q$y[j]),(p$x[i]-p$x[j]),(q$x[j]-q$x[i]),(p$y[j]-p$y[i])),nrow=1))
  BB=rbind(BB,matrix(p$x[i]*q$y[i]-p$x[j]*q$y[j]-p$y[i]*q$x[i]+p$y[j]*q$x[j]))
}
s=solve(AA,BB)
X0=s[1]
Y0=s[3]
A=s[4]
B=s[2]

EE=matrix(c(1,1,(X0-p$x[1])/(q$x[1]-A),(X0-p$x[2])/(q$x[2]-A)),ncol=2)
FF=c(p$z[1]+q$z[1]*(X0-p$x[1])/(q$x[1]-A),p$z[2]+q$z[2]*(X0-p$x[2])/(q$x[2]-A))

Z0=solve(EE,FF)[1]
C=solve(EE,FF)[2]

format(X0+Y0+Z0,scientific=F)

**Reply by EllaKaye** [reply]
Impressive! Thank you for sharing. Yes, do sign up for the leaderboard next year!
